### ISSUE TITLE
Slow queries with ORC format

### ISSUE THREAD

**Issue by stagraqubole** [facebook/presto#4160]
Seeing this consistently with ORC table. Trying queries of form "select * from TABLE order by COL1", the table is very small (~20MB data, 500k rows). Query took 40min to complete on a 2-node cluster.
Same query against TXT finishes in around a minute.

The problem seems to be happening due to ORC Reader using common Dictionary block. In my case, the Dictionary block read had 500k entries. Since the same block was shared amongst all pages, page.getSizeinBytes() returned size based on the shared block which comes to be ~20MB.

PageSplitter.splitPage tries to bring down the page size but gives up eventually after splitting pages down to one position per page. Since the dictionary block is shared hence even after splitting the size does not come down by much. Finally during serialization, compact dictionary is formed and the data transfer between TableScan and OrderBy operators happens one position at a time. This takes lots of round trips to transfer all data and hence performance degrades.

To verify this, I changed DictionaryBlock to return size considering only the entries in Dictionary that will be used by it. With that, the times came down to normal.

**Comment by stagraqubole** [comment]
To hot fix, this is what I did:
Changed:
```
public DictionaryBlock(int positionCount, Block dictionary, Slice ids)
    {
        ....
        this.sizeInBytes = dictionary.getRetainedSizeInBytes() + ids.getRetainedSize();
    }
```
To:
```
public DictionaryBlock(int positionCount, Block dictionary, Slice ids)
    {
        ....
        int referencedDictionarySize = 0;
        for (int i = 0; i < positionCount; i++) {
            referencedDictionarySize += dictionary.getLength(getIndex(i));
        }

        this.sizeInBytes = referencedDictionarySize + ids.getRetainedSize();
    }
```

I can open the PR for this but I am not sure why is is full dictionary size considered for the page Size originally. Waiting to get clarification

**Comment by haozhun** [comment]
@stagraqubole This is really good find. Thank you

1. You are correct that `getSizeInBytes` is wrong in `DictionaryBlock`. But we did it consciously because the alternative would be the code you presented above, which is too costly in our opinion. We should probably add a comment.
2. We believe we can fix your issue by fixing two bugs:
 * `PageSplitterUtil.splitPage` should call `copyRegion` instead of `getRegion`.
 * Fix implementation of `ArrayBlock.copyRegion`, whose behavior violates the contract of `Block`. It is supposed to copy and block. But it didn't.

**Comment by haozhun** [comment]
@stagraqubole You are welcome to submit a pull request for this. Otherwise, @nileema will fix this when she becomes less busy.

**Comment by stagraqubole** [comment]
Thanks @haozhun, few questions:
1. Keeping getSizeInBytes same and doing copyRegion to compact Dictionary would still divide the page into two even when one page could have been transferred. This might not be a big issue, two pages rather than 1000 is still better.
2. Wouldn't changing splitPage to use copyRegion have performance impact?

**Comment by nileema** [comment]
We decided to go with the approach of fixing getSizeInBytes for dictionary blocks. The sizeInBytes is actually the sum of lengths of all referenced elements of the dictionary, rather than the unrolled dictionary. 

Also, It is ok for the pageSplitter to call ``getRegion`` because the ``DictionaryBlockEncoding`` always calls compact before it sends the data.  

PR: #4191

**Comment by stagraqubole** [comment]
Great, I will pull that in. Thanks.

**Comment by dain** [comment]
This was fixed by #4191
